snapshotter/demux/proxy/snapshotter.go

```diff
@@ -28,8 +30,31 @@ import (
 // SnapshotterDialer defines an interface for establishing a network connection.
 type SnapshotterDialer = func(context.Context, *logrus.Entry, string, uint32)
 
+// RemoteSnapshotter embeds a snapshots.Snapshotter and its metrics proxy.
+type RemoteSnapshotter struct {
+	snapshots.Snapshotter
+	MetricsProxy *metrics.Proxy
+}
```

Comment:
Do we need to expose these types? Could they be remoteSnapshotter and metricsProxy?

Reply:
I can refactor to allow metricsProxy to not be exposed. 

As for exposing RemoteSnapshotter as a type, I think the argument for exposing would be that if this were not exposed, then there would have to be some type assertions in other areas. E.g. In `snapshotter/app/service.go` , `snapshotter/demux/cache/snapshotter_cache.go`, etc. where we would change the signatures of many of these functions to be of type `snapshots.Snapshotter` and assert that the returned valued is a `proxy.remoteSnapshotter` before performing any remote specific operations.

Reply:
Regarding type assertions, you could potentially do something like this:

```
type RemoteSnapshotter interface {
    snapshots.Snapshotter    // Extend the snapshots.Snapshotter interface.
    NewProxySnapshotter( ... )
}

// Private struct.  Implements RemoteSnapshotter and Snapshotter interfaces.
type remoteSnapshotter struct {
	snapshots.Snapshotter
	metricsProxy *metrics.Proxy
}

// NewProxySnapshotter creates a proxy snapshotter using gRPC over vsock connection.
func NewProxySnapshotter( ... ) (RemoteSnapshotter, error) {    // Now returns your implementation of the interface.
     ...
}
```

You'll still need some refactoring to expose whichever aspects of `MetricsProxy` that you're currently using elsewhere, but if you _did_ want to make this private, this'd probably be the best approach.

Reply:
What are the remote snapshotter-specific operations?

Generally speaking, it is better to have fewer, larger packages.

https://dave.cheney.net/practical-go/presentations/qcon-china.html#_consider_fewer_larger_packages

Reply:
The remote snapshotter doesn't really have specific operations that aren't already defined by the `snapshots.Snapshotter` type... the main caveat being that on close, the remote snapshotter will also shut down the metrics proxy server.